glowroot-central start up fails with error - gc_grace_seconds value must be zero

Hi,

Trying to configure glowroot-central-0.14.0-beta with CosmosDB Cassandra API. Facing below error while startup, even though my 

- glowroot-central.properties reads,

##default is 4h. For CosmosDB Cassandra API, set this value to 0.
**cassandra.gcGraceSeconds=0**

- verified that in cassandra, the value of gc_grace_seconds is also 0 for glowroot keyspace.

Command used to run glowroot:
java -jar glowroot-central.jar

**Error:**
2022-04-08 05:11:52.838 INFO  org.glowroot - Glowroot version: 0.14.0-beta, built 2021-10-26 00:33:26 +0000
2022-04-08 05:11:52.842 INFO  org.glowroot - Java version: 1.8.0_312
2022-04-08 05:11:53.081 INFO  c.datastax.driver.core.ClockFactory - Using native clock to generate timestamps.
2022-04-08 05:11:53.089 INFO  com.datastax.driver.core - DataStax Java driver 3.7.2 for Apache Cassandra
2022-04-08 05:11:53.095 INFO  c.d.driver.core.GuavaCompatibility - Detected Guava >= 19 in the classpath, using modern compatibility layer
2022-04-08 05:11:53.488 INFO  com.datastax.driver.core.NettyUtil - Did not find Netty's native epoll transport in the classpath, defaulting to NIO.
2022-04-08 05:11:54.570 INFO  c.d.d.c.p.DCAwareRoundRobinPolicy - Using data-center name '******' for DCAwareRoundRobinPolicy (if this is incorrect, please provide the correct datacenter name with DCAwareRoundRobinPolicy constructor)
2022-04-08 05:11:54.572 INFO  com.datastax.driver.core.Cluster - New Cassandra host xxxxx/xxxxxx:10350 added
2022-04-08 05:11:54.888 INFO  org.glowroot - connected to Cassandra (version 3.11.0), using keyspace 'glowroot' (replication factor 1) and consistency level QUORUM
2022-04-08 05:11:55.203 INFO  org.glowroot - creating glowroot central schema...
2022-04-08 05:11:55.462 INFO  c.d.driver.core.RequestHandler - Query select value from central_config where key = ? is not prepared on xxxxx/xxxxx:10350, preparing before retrying executing. Seeing this message a few times is fine, but seeing it a lot may be source of performance problems
2022-04-08 05:11:55.649 **ERROR **org.glowroot - gc_grace_seconds value must be zero.****
com.datastax.driver.core.exceptions.InvalidQueryException: gc_grace_seconds value must be zero.
        at com.datastax.driver.core.exceptions.InvalidQueryException.copy(InvalidQueryException.java:49)
        at com.datastax.driver.core.DriverThrowables.propagateCause(DriverThrowables.java:35)
        at com.datastax.driver.core.DefaultResultSetFuture.getUninterruptibly(DefaultResultSetFuture.java:293)
        at com.datastax.driver.core.AbstractSession.execute(AbstractSession.java:58)
        at com.datastax.driver.core.AbstractSession.execute(AbstractSession.java:39)
        at org.glowroot.central.util.Session.updateSchemaWithRetry(Session.java:440)
        at org.glowroot.central.util.Session.updateSchemaWithRetry(Session.java:353)
        at org.glowroot.central.util.Session.createTableWithTracking(Session.java:368)
        at org.glowroot.central.util.Session.createTableWithTWCS(Session.java:305)
        at org.glowroot.central.util.Session.createTableWithTWCS(Session.java:285)
        at org.glowroot.central.util.Session.createTableWithTWCS(Session.java:280)
        at org.glowroot.central.repo.ActiveAgentDao.<init>(ActiveAgentDao.java:92)
        at org.glowroot.central.repo.CentralRepoModule.<init>(CentralRepoModule.java:86)
        at org.glowroot.central.CentralModule.<init>(CentralModule.java:201)
        at org.glowroot.central.CentralModule.create(CentralModule.java:137)
        at org.glowroot.central.CentralModule.create(CentralModule.java:132)
        at org.glowroot.central.Main.start(Main.java:44)
        at org.glowroot.central.Main.main(Main.java:28)
**Caused by: com.datastax.driver.core.exceptions.InvalidQueryException: gc_grace_seconds value must be zero.**
        at com.datastax.driver.core.Responses$Error.asException(Responses.java:181)
        at com.datastax.driver.core.DefaultResultSetFuture.onSet(DefaultResultSetFuture.java:215)
        at com.datastax.driver.core.RequestHandler.setFinalResult(RequestHandler.java:236)
        at com.datastax.driver.core.RequestHandler.access$2600(RequestHandler.java:62)
        at com.datastax.driver.core.RequestHandler$SpeculativeExecution.setFinalResult(RequestHandler.java:1005)
        at com.datastax.driver.core.RequestHandler$SpeculativeExecution.onSet(RequestHandler.java:808)
        at com.datastax.driver.core.Connection$Dispatcher.channelRead0(Connection.java:1240)
        at com.datastax.driver.core.Connection$Dispatcher.channelRead0(Connection.java:1158)
        at io.netty.channel.SimpleChannelInboundHandler.channelRead(SimpleChannelInboundHandler.java:99)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:379)
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:365)
      .....
**Exception in thread "main" com.datastax.driver.core.exceptions.InvalidQueryException: gc_grace_seconds value must be zero.**

--- a/central/src/main/java/org/glowroot/central/CentralModule.java
+++ b/central/src/main/java/org/glowroot/central/CentralModule.java
@@ -667,6 +667,11 @@ public class CentralModule {
         if (!Strings.isNullOrEmpty(cassandraMaxConcurrentQueries)) {
             builder.cassandraMaxConcurrentQueries(Integer.parseInt(cassandraMaxConcurrentQueries));
         }
+        String cassandraGcGraceSeconds =
+                properties.get("glowroot.cassandra.gcGraceSeconds");
+        if (!Strings.isNullOrEmpty(cassandraGcGraceSeconds)) {
+            builder.cassandraGcGraceSeconds(Integer.parseInt(cassandraGcGraceSeconds));
+        }
         String cassandraPoolTimeoutMillis = properties.get("glowroot.cassandra.pool.timeoutMillis");
         if (!Strings.isNullOrEmpty(cassandraPoolTimeoutMillis)) {
             builder.cassandraPoolTimeoutMillis(Integer.parseInt(cassandraPoolTimeoutMillis));

--- a/central/src/main/webapp/META-INF/glowroot-central.properties
+++ b/central/src/main/webapp/META-INF/glowroot-central.properties
@@ -19,6 +19,9 @@ cassandra.consistencyLevel=
 # default is cassandra.maxConcurrentQueries=1024
 cassandra.maxConcurrentQueries=
 
+# default is 4h. For CosmosDB Cassandra API, set this value to 0.
+cassandra.gcGraceSeconds=
+
 # default is grpc.bindAddress=0.0.0.0
 grpc.bindAddress=
 

